### DISCUSSION TITLE
Parsec - how to implement recursive grammar?

### DISCUSSION THREAD

**Discussion by AlesSturala** [louthy/language-ext#1004] [Q&A] [answered]
Hello,

I hope that someone can help me. I'm trying to implement a simple (but recursive) grammar:

E = ( E ) | term

But I cannot figure out how to recursively call itself: E = ( E )

I tried the following but it fails on _NullPointerException_:

`            Parser<Term> ex = null;

            ex = from ws in choice(between(lb, rb, ex), number)
                     select ws;

            var res = parse(ex, "( 10 )");
            WriteLine(res.ToString());`

Any help is appreciated, thank you in advance.

**Comment by louthy** [comment, marked as answer]
So, here's a few tips:

## 1. Declare `ex` before-hand as `null` 

```c#
   Parser<YourExpr> ex = null;
```

## 2. The first line of the LINQ expression can't refer to `ex`

```c#
    ex = from ... in THIS_WILL_BE_EVALUATED_IMMEDIATELY
         from ... in THIS_WILL_BE_EVALUATED_LAZILY
         select ...;
```

Because the right-hand-side of the first `in` is evaluated immediately, `ex` will be `null`.  And so, your `between` parser won't be particularly useful.

You can fix this (in general), by using `lazyp(() => ex)`, which will evaluate `ex` later, when it's got a value.  

```c#
    ex = from ws in choice(between(lb, rb, lazyp(() => ex)), number)
         select ws;
```

Or, move the recursion to the second line of the parser, this can be done by consuming the `unitp` parser:

```c#
    ex = from _ in unitp
         from ws in choice(between(lb, rb, ex), number)
         select ws;
```

## 3. Take a look at some examples:

The [AccountingDSL sample](https://github.com/louthy/language-ext/tree/main/Samples/AccountingDSL/AccountingDSL) has a [script-parser](https://github.com/louthy/language-ext/blob/main/Samples/AccountingDSL/AccountingDSL/ScriptParsing/ScriptParser.cs), which can parse scripts like this:

```c#
TotalAssets = sum(filter("Asset", rows));
TotalExpenses = sum(filter("Expense", rows));
Balance = TotalAssets - TotalExpenses;
log("Total Assets : {TotalAssets}");
log("Total Expenses : {TotalExpenses}");
log("Balance (Total Assets - Total Expenses): {Balance}");
```
